src/dataflow-types/src/types.rs

```diff
@@ -1307,42 +1313,115 @@ pub mod sources {
-            use aws_types::region::Region;
-
-            let mut loader = aws_config::from_env().region(Region::new(self.region.clone()));
-            if let Some(credentials) = &self.credentials {
-                loader = loader.credentials_provider(SharedCredentialsProvider::new(
-                    aws_types::Credentials::from_keys(
-                        &credentials.access_key_id,
-                        &credentials.secret_access_key,
-                        credentials.session_token.clone(),
-                    ),
-                ));
-            }
+            let mut loader: aws_config::ConfigLoader = aws_config::from_env();
+            // RegionProviderChain is not Clone so we must construct a new on for each use
+            let region = || {
+                if let Some(region) = &self.region {
+                    RegionProviderChain::first_try(Region::new(region.clone()))
+                        .or_default_provider()
+                } else {
+                    let mut wc =
+                        aws_config::default_provider::region::DefaultRegionChain::builder();
+                    if let AwsCredentials::Profile { profile_name } = &self.credentials {
+                        wc = wc.profile_name(profile_name);
+                    }
+                    // TODO: this doesn't work for profiles which do not provide their own region
+                    // but instead specify a source_profile that does have a region
+                    RegionProviderChain::first_try(wc.build()).or_default_provider()
+                }
+            };
+
+            loader = loader.region(region());
+
+            let cred_provider = match &self.credentials {
+                AwsCredentials::Default => SharedCredentialsProvider::new(
+                    DefaultCredentialsChain::builder()
+                        .region(region())
+                        .build()
+                        .await,
+                ),
+                AwsCredentials::Profile { profile_name } => SharedCredentialsProvider::new(
+                    DefaultCredentialsChain::builder()
+                        .profile_name(profile_name)
+                        .region(region())
+                        .build()
+                        .await,
+                ),
+                AwsCredentials::Static {
+                    access_key_id,
+                    secret_access_key,
+                    session_token,
+                } => SharedCredentialsProvider::new(aws_types::Credentials::from_keys(
+                    access_key_id,
+                    secret_access_key,
+                    session_token.clone(),
+                )),
+            };
+
+            let role = if let Some(role_arn) = self.role_arn.as_ref() {
+                let mut role = AssumeRoleProvider::builder(role_arn);
+
+                // this affects which region to perform STS on, not where anything else happens
```

Comment:
Does it matter what region we perform STS on? Would folks ever want to perform STS in a different region than the one they want to perform other API operations on?

Reply:
Right, the point of of this is that we are setting STS to use the same region as everything else. The comment might  be useless, but the exact connection between the various things in the various providers is subtle enough that it seemed worth calling out.

Reply:
Sorry, not complaining about the comment! Those were legitimate questions. Like, is there a region we're _supposed_ to be performing the STS AssumeRole operation on; it is sometimes distinct from the region the customer has specified in the source (e.g., you always AssumeRole in the region that is running `materialized` even if you are connecting to an S3 bucket in a different region); would the customer ever want/need to be able to configure the STS region separately.